### PR TITLE
cephfs-journal-tool: Add preventive measures to avoid fs corruption

### PR DESCRIPTION
**Pull request by joscollin**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/cephfs/JournalTool.cc [hunk 1]
```diff
@@ -250,17 +257,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
     }
   } else if (command == "reset") {
     bool force = false;
-    if (argv.size() == 2) {
+    if (argv.size() == 1) {
+        std::cerr << "warning: this operation resets the journal!!!\n"
```
Comment (gregsfortytwo) [inline src/tools/cephfs/JournalTool.cc:261]: These messages should be more explicit. Something like "Do not run this operation if you do not understand CephFS' internal storage mechanisms or have received specific instructions from those who do"

File: src/tools/cephfs/JournalTool.cc [hunk 2]
```diff
@@ -250,17 +257,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
...
+      }
+    } else if (argv.size() == 3) {
       if (std::string(argv[1]) == "--force") {
         force = true;
       } else {
         std::cerr << "Unknown argument " << argv[1] << std::endl;
         return -EINVAL;
       }
-    } else if (argv.size() > 2) {
+      if (std::string(argv[2]) == "" || std::string(argv[2]) != "--yes-i-really-really-mean-it") {
```
Comment (chrisphoffman) [inline src/tools/cephfs/JournalTool.cc:279]: What does checking for empty string do here? Do we need it for `argv.size() == 2`?

File: src/tools/cephfs/JournalTool.cc [hunk 3]
```diff
@@ -250,17 +257,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
...
+        std::cerr << "warning: this operation resets the journal!!!\n"
+                  << "         if you want to continue, please add --yes-i-really-really-mean-it!!!"
+                  << std::endl;
+        return -EINVAL;
+      }
+    } else if (argv.size() > 3) {
       std::cerr << "Too many arguments!" << std::endl;
       return -EINVAL;
     }
+
```
Comment (chrisphoffman) [inline src/tools/cephfs/JournalTool.cc:289]: remove new line

File: src/tools/cephfs/JournalTool.cc (hunk 1 above)
Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc:261]: fixed

File: src/tools/cephfs/JournalTool.cc (hunk 2 above)
Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc]: Yes, that's unnecessary check. Fixed it.

**Comment by joscollin** [comment]
jenkins test make check

File: src/tools/cephfs/JournalTool.cc [hunk 4]
```diff
@@ -139,6 +139,13 @@ int JournalTool::main(std::vector<const char*> &argv)
   }
  
   auto& fs = fsmap->get_filesystem(role_selector.get_ns());
+  int rnk;
```
Comment (vshankar) [inline src/tools/cephfs/JournalTool.cc:142]: nit: just use `rank`.

File: src/tools/cephfs/JournalTool.cc [hunk 5]
```diff
@@ -139,6 +139,13 @@ int JournalTool::main(std::vector<const char*> &argv)
...
+  stringstream (rank_str.substr(rank_str.find(':') + 1)) >> rnk;
+  if (fs.get_mds_map().is_active(rnk)) {
+    derr << "Cannot run cephfs-journal-tool on an active Filesystem!" << dendl;
```
Comment (vshankar) [inline src/tools/cephfs/JournalTool.cc:145]: nit: file system.

File: src/tools/cephfs/JournalTool.cc [hunk 6]
```diff
@@ -250,14 +250,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
...
+        return -EINVAL;
+    } else if (argv.size() == 2) {
+      if (std::string(argv[1]) == "--force") {
+        std::cerr << "warning: this operation resets the journal!!!\n"
+                  << "Do not run this operation if you do not understand CephFS' internal storage mechanisms or have received specific instructions from those who do.\n"
+                  << "If you want to continue, please add --yes-i-really-really-mean-it!!!"
+                  << std::endl;
+        return -EINVAL;
+      } else if (std::string(argv[1]) != "--yes-i-really-really-mean-it") {
+        std::cerr << "Unknown argument " << argv[1] << std::endl;
```
Comment (vshankar) [inline src/tools/cephfs/JournalTool.cc:274]: So, `--force` is now replaced by `--yes-i-really-really-mean-it`?

**Review by vshankar — CHANGES_REQUESTED** [review]
This change would require mentions in documentation and PendingReleaseNotes.

File: src/tools/cephfs/JournalTool.cc (hunk 6 above)
Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc:273]: No, `--force` i there. But we need to use `--yes-i-really-really-mean-it` for `journal reset --force` command to continue it's execution.

**Comment by joscollin** [comment]
@vshankar Will update the documentation.

File: src/tools/cephfs/JournalTool.cc (hunk 6 above)
Comment (vshankar) [reply src/tools/cephfs/JournalTool.cc:273]: ok, `--force` is actually a hard reset.

**Comment by joscollin** [comment]
@vshankar Updated the documentation. Please check.

File: PendingReleaseNotes [hunk 7]
```diff
@@ -148,6 +148,8 @@ CephFS: Disallow delegating preallocated inode ranges to clients. Config
    `ceph config get mgr mgr/volumes/snapshot_clone_no_wait`
   and it can be disabled by using:
    `ceph config set mgr mgr/volumes/snapshot_clone_no_wait false`
+* CephFS: cephfs-journal-tool prevent itself from working on an active filesystem.
```
Comment (vshankar) [inline PendingReleaseNotes:151]: reword this as: cephfs-journal-tool is guarded against running on an online file system.

File: PendingReleaseNotes [hunk 8]
```diff
@@ -148,6 +148,8 @@ CephFS: Disallow delegating preallocated inode ranges to clients. Config
...
+  The 'journal reset' command works only if --yes-i-really-really-mean-it is added.
```
Comment (vshankar) [inline PendingReleaseNotes:152]: Use full "journal reset" command and reword as

.... requires <flag>.

**Review by vshankar — CHANGES_REQUESTED** [review]
We might include/change qa tests that needs fixing. Please check those.

**Review by chrisphoffman — COMMENTED** [review]
You missed adding `--yes-i-really-really-mean-it` to one instance of `cephfs-journal-tool journal reset`

see:
https://github.com/ceph/ceph/blob/2389f0054e98a4aad7f3e1b36077bc07ea6d955f/src/tools/cephfs/first-damage.py#L28

**Comment by joscollin** [comment]
> You missed adding `--yes-i-really-really-mean-it` to one instance of `cephfs-journal-tool journal reset`
> 
> see:
> 
> https://github.com/ceph/ceph/blob/2389f0054e98a4aad7f3e1b36077bc07ea6d955f/src/tools/cephfs/first-damage.py#L28

Fixed

**Comment by joscollin** [comment]
@vshankar @chrisphoffman ping

File: src/tools/cephfs/JournalTool.cc [hunk 9]
```diff
@@ -250,14 +250,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
     }
   } else if (command == "reset") {
     bool force = false;
-    if (argv.size() == 2) {
+    if (argv.size() == 1) {
```
Comment (vshankar) [inline src/tools/cephfs/JournalTool.cc:259]: wait.. shouldn't this throw up help?

**Review by vshankar — CHANGES_REQUESTED** [review]
@joscollin This would require a rebase since more invocations of the journal tool were recently merged.

```
qa/tasks/cephfs/test_journal_migration.py:        # Verify that cephfs-journal-tool can now read the rewritten journal
qa/tasks/vstart_runner.py:    require_binaries = ["ceph-dencoder", "cephfs-journal-tool", "cephfs-data-scan",
qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
qa/workunits/fs/damage/test-first-damage.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
qa/workunits/fs/damage/test-first-damage.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
qa/workunits/suites/cephfs_journal_tool_smoke.sh:export BIN="${BIN:-cephfs-journal-tool --rank=cephfs:0}"
```

File: src/tools/cephfs/JournalTool.cc (hunk 9 above)
Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc:259]: No, because [std::string command = argv[0]](https://github.com/ceph/ceph/blob/main/src/tools/cephfs/JournalTool.cc#L237). At this point, when `argv.[0] == "reset"`, we've already got the command. So no need of throwing help here. Just issue a warning for adding the option.

Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc:259]: EDIT: I meant https://github.com/ceph/ceph/blob/main/src/tools/cephfs/JournalTool.cc#L231

**Comment by joscollin** [comment]
> @joscollin This would require a rebase since more invocations of the journal tool were recently merged.
> 
> ```
> qa/tasks/cephfs/test_journal_migration.py:        # Verify that cephfs-journal-tool can now read the rewritten journal
> qa/tasks/vstart_runner.py:    require_binaries = ["ceph-dencoder", "cephfs-journal-tool", "cephfs-data-scan",
> qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
> qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
> qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
> qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
> qa/workunits/fs/damage/test-first-damage-lost-found.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
> qa/workunits/fs/damage/test-first-damage.sh:  cephfs-journal-tool --rank="$FS":0 event recover_dentries summary
> qa/workunits/fs/damage/test-first-damage.sh:  cephfs-journal-tool --rank="$FS":0 journal reset
> qa/workunits/suites/cephfs_journal_tool_smoke.sh:export BIN="${BIN:-cephfs-journal-tool --rank=cephfs:0}"
> ```

@vshankar Rebased.

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
@vshankar Please review again.

**Comment by vshankar** [comment]
> @vshankar Please review again.

Sorry, why? What changed? I had reviewed the last revision after you pushed and then I had to go on PTO.

File: src/tools/cephfs/JournalTool.cc [hunk 10]
```diff
@@ -139,6 +139,12 @@ int JournalTool::main(std::vector<const char*> &argv)
   }
  
   auto& fs = fsmap->get_filesystem(role_selector.get_ns());
+  stringstream (rank_str.substr(rank_str.find(':') + 1)) >> rank;
+  if (fs.get_mds_map().is_active(rank)) {
+    derr << "Cannot run cephfs-journal-tool on an active file system!" << dendl;
+    return -CEPHFS_EPERM;
+  }
```
Comment (dparmar18) [inline src/tools/cephfs/JournalTool.cc:146]: so we have completely disabled running journal tool on an active fs, it is fine since it will not cause additional damage but what if the admin wants to really do it since he/she knows what he/she is upto? is it always deadly to run this tool on active fs or can it be run but with extreme caution?

Comment (vshankar) [reply src/tools/cephfs/JournalTool.cc:146]: > so we have completely disabled running journal tool on an active fs, it is fine since it will not cause additional damage but what if the admin wants to really do it since he/she knows what he/she is upto?

I'l be really afraid of such cluster admins :)

If they *really really* want to, then they can compile their own version of the patched up tool and run at will. I don't see a case where we should allow the tool to run on an active fs. If someone can show me a case, I'll be happy to change my mind.

Comment (dparmar18) [reply src/tools/cephfs/JournalTool.cc:146]: i don't know of any such case, maybe it is fine to go this way and change back in future if a case arises

File: qa/tasks/cephfs/test_data_scan.py [hunk 11]
```diff
@@ -447,9 +447,9 @@ def get_state(mds_id):
         if False:
             with self.assertRaises(CommandFailedError):
                 # Normal reset should fail when no objects are present, we'll use --force instead
-                self.fs.journal_tool(["journal", "reset"], 0)
+                self.fs.journal_tool(["journal", "reset", "--yes-i-really-really-mean-it"], 0)
 
-        self.fs.journal_tool(["journal", "reset", "--force"], 0)
+        self.fs.journal_tool(["journal", "reset", "--force", "--yes-i-really-really-mean-it"], 0)
```
Comment (dparmar18) [inline qa/tasks/cephfs/test_data_scan.py:452]: it seems journal reset is widely used in test cases, how about creating a new func `journal_reset()`?

Comment (joscollin) [reply qa/tasks/cephfs/test_data_scan.py:452]: @dparmar18 
Could you please create a tracker for that (and assign to me), so that this `urgent` patch could make progress on it's own ?

**Review by dparmar18 — APPROVED** [review]
LGTM

**Review by vshankar — CHANGES_REQUESTED** [review]
@joscollin I think there are a few more instance in tests to fix. See: https://pulpito.ceph.com/vshankar-2024-05-01_17:34:00-fs-wip-vshankar-testing-20240430.111407-debug-testing-default-smithi/7683778/

```
2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:======================================================================
2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:ERROR: test_journal_migration (tasks.cephfs.test_journal_migration.TestJournalMigration)
2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/test_journal_migration.py", line 70, in test_journal_migration
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    inspect_out = self.fs.journal_tool(["journal", "inspect"], 0)
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1715, in journal_tool
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    return self._run_tool("cephfs-journal-tool", args, fs_rank, quiet)
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1694, in _run_tool
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self.tool_remote.sh(script=base_args + args, stdout=StringIO()).strip()
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 97, in sh
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    proc = self.run(**kwargs)
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 523, in run
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 455, in run
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r.wait()
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 161, in wait
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 181, in _raise_for_status
2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:    raise CommandFailedError(
2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi016 with status 255: 'cephfs-journal-tool --debug-mds=20 --debug-ms=1 --debug-objecter=1 --rank cephfs:0 journal inspect'
```

**Comment by joscollin** [comment]
> @joscollin I think there are a few more instance in tests to fix. See: https://pulpito.ceph.com/vshankar-2024-05-01_17:34:00-fs-wip-vshankar-testing-20240430.111407-debug-testing-default-smithi/7683778/
> 
> ```
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:======================================================================
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:ERROR: test_journal_migration (tasks.cephfs.test_journal_migration.TestJournalMigration)
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/test_journal_migration.py", line 70, in test_journal_migration
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    inspect_out = self.fs.journal_tool(["journal", "inspect"], 0)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1715, in journal_tool
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    return self._run_tool("cephfs-journal-tool", args, fs_rank, quiet)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1694, in _run_tool
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self.tool_remote.sh(script=base_args + args, stdout=StringIO()).strip()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 97, in sh
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    proc = self.run(**kwargs)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 523, in run
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 455, in run
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r.wait()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 161, in wait
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 181, in _raise_for_status
> 2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:    raise CommandFailedError(
> 2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi016 with status 255: 'cephfs-journal-tool --debug-mds=20 --debug-ms=1 --debug-objecter=1 --rank cephfs:0 journal inspect'
> ```

@vshankar 
`Cannot run cephfs-journal-tool on an active file system!` Those tests were running `cephfs-journal-tool` on an active filesystem. It should do an `fs fail <fsname>`, before running `cephfs-journal-tool`, if there's an active filesystem.

**Comment by joscollin** [comment]
> @joscollin I think there are a few more instance in tests to fix. See: https://pulpito.ceph.com/vshankar-2024-05-01_17:34:00-fs-wip-vshankar-testing-20240430.111407-debug-testing-default-smithi/7683778/
> 
> ```
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:======================================================================
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:ERROR: test_journal_migration (tasks.cephfs.test_journal_migration.TestJournalMigration)
> 2024-05-01T22:10:13.292 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/test_journal_migration.py", line 70, in test_journal_migration
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    inspect_out = self.fs.journal_tool(["journal", "inspect"], 0)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1715, in journal_tool
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    return self._run_tool("cephfs-journal-tool", args, fs_rank, quiet)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5ef2baf9bfba0061cb137abe5d8fc74e1acd4bd8/qa/tasks/cephfs/filesystem.py", line 1694, in _run_tool
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self.tool_remote.sh(script=base_args + args, stdout=StringIO()).strip()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 97, in sh
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    proc = self.run(**kwargs)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/remote.py", line 523, in run
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 455, in run
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    r.wait()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 161, in wait
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
> 2024-05-01T22:10:13.293 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_teuthology_1ae7ad82388e92a475afff437d49054826c019a1/teuthology/orchestra/run.py", line 181, in _raise_for_status
> 2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:    raise CommandFailedError(
> 2024-05-01T22:10:13.294 INFO:tasks.cephfs_test_runner:teuthology.exceptions.CommandFailedError: Command failed on smithi016 with status 255: 'cephfs-journal-tool --debug-mds=20 --debug-ms=1 --debug-objecter=1 --rank cephfs:0 journal inspect'
> ```

This is fixed and the changes are pushed. There's a similar failure with test_flush.py, which is WIP.

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by joscollin** [comment]
@vshankar The qa failures are fixed. This PR is ready for another review and qa.

**Comment by vshankar** [comment]
@joscollin Could you update here how the assert in the tests was fixed?

**Comment by joscollin** [comment]
> @joscollin Could you update here how the assert in the tests was fixed?

The AssertionError [1] was not because of  string mismatch. When we do `fs.fail()`, https://github.com/ceph/ceph/blob/main/qa/tasks/cephfs/test_flush.py#L73 needs an active fs. That was the issue. I've fixed it by making the fs active/inactive at the right places: https://github.com/ceph/ceph/pull/55758/commits/0820b31d5b1f1542636c56611ec27636afc23b68

[1] https://qa-proxy.ceph.com/teuthology/jcollin-2024-05-08_05:08:43-fs:functional-wip-jcollin-testing-07052024-distro-default-smithi/7697306/teuthology.log

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66035](https://tracker.ceph.com/issues/66035).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66063](https://tracker.ceph.com/issues/66063).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66090](https://tracker.ceph.com/issues/66090).

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2024-05-28
